File: tests/e2e/javascript/eventhub.ts
import Device from '../../../sdk/javascript/Device';
import Sdk from '../../../sdk/javascript/Sdk';
import Service from '../../../sdk/javascript/Service';
import ServiceRequest from '../../../sdk/javascript/ServiceRequest';
import ServiceResponse from '../../../sdk/javascript/ServiceResponse';
import { fatal, getCredentials, log } from './util';

const ORG_ID = 'Org2MSP';
const ORG_DOMAIN = 'org2.example.com';
const USER_NAME = 'User1@org2.example.com';
const PEER_NAME = 'peer0.org2.example.com';
const PEER_ENDPOINT = 'localhost:9051';

async function handleDeviceEvents(isb: Sdk) {
  log('Watching for device events');

  const events = await isb.getDeviceRegistry().registerEvent();
  const expected: { [action: string]: number } = {
    register: 1,
    deregister: 1,
  };
  const actual: { [action: string]: number } = {};

  outer: for await (const event of events) {
    actual[event.action] = (actual[event.action] ?? 0) + 1;

    if (event.action === 'register' || event.action === 'deregister') {
      const device = event.payload as Device;
      if (device.id !== event.deviceId || device.organizationId !== event.organizationId) {
        fatal('device ID or organization ID mismatch');
      }
    }

    for (const [action, value] of Object.entries(expected)) {
      if (actual[action] !== value) {
        continue outer;
      }
    }

    break;
  }

  for (const [action, value] of Object.entries(expected)) {
    if (actual[action] !== value) {
      fatal(`should have received ${value} device ${action} events`);
    }
  }

  events.close();
  log('Done watching for device events');
}

async function handleServiceEvents(isb: Sdk) {
  log('Watching for service events');

  const events = await isb.getServiceRegistry().registerEvent();
  const expected: { [action: string]: number } = {
    register: 2,
    deregister: 1,
  };
  const actual: { [action: string]: number } = {};

  outer: for await (const event of events) {
    actual[event.action] = (actual[event.action] ?? 0) + 1;

    if (event.action === 'register' || event.action === 'deregister') {
      const service = event.payload as Service;
      if (
        service.deviceId !== event.deviceId ||
        service.organizationId !== event.organizationId ||
        service.name !== event.serviceName
      ) {
        fatal('event and payload device ID, organization ID, or service name mismatch');
      }
    }

    for (const [action, value] of Object.entries(expected)) {
      if (actual[action] !== value) {
        continue outer;
      }
    }

    break;
  }

  for (const [action, value] of Object.entries(expected)) {
    if (actual[action] !== value) {
      fatal(`should have received ${value} service ${action} events`);
    }
  }

  events.close();
  log('Done watching for service events');
}

async function handleRequestEvents(isb: Sdk) {
  log('Watching for service request events');

  const events = await isb.getServiceBroker().registerEvent();
  const expected: { [action: string]: number } = {
    request: 2,
    respond: 2,
    remove: 1,
  };
  const actual: { [action: string]: number } = {};

  outer: for await (const event of events) {
    actual[event.action] = (actual[event.action] ?? 0) + 1;

    if (event.action === 'request') {
      const request = event.payload as ServiceRequest;
      if (request.id !== event.requestId) {
        fatal('event and payload request ID mismatch');
      }
    } else if (event.action === 'respond') {
      const response = event.payload as ServiceResponse;
      if (response.requestId !== event.requestId) {
        fatal('event and payload request ID mismatch');
      }
    } else if (event.action === 'remove') {
      if (event.payload !== event.requestId) {
        fatal('event and payload request ID mismatch');
      }
    }

    for (const [action, value] of Object.entries(expected)) {
      if (actual[action] !== value) {
        continue outer;
      }
    }

    break;
  }

  for (const [action, value] of Object.entries(expected)) {
    if (actual[action] !== value) {
      fatal(`should have received ${value} service request ${action} events`);
    }
  }

  events.close();
  log('Done watching for service request events');
}

async function main() {
  const [certificate, privateKey, tlsCertificate] = getCredentials(
    process.env.FABRIC_ROOT ?? '.',
    ORG_DOMAIN,
    USER_NAME,
    PEER_NAME,
  );

  const isb = new Sdk({
    organizationId: ORG_ID,
    certificate,
    privateKey,
    gatewayPeerEndpoint: PEER_ENDPOINT,
    gatewayPeerServerName: PEER_NAME,
    gatewayPeerTLSCertificate: tlsCertificate,
    networkName: 'mychannel',
    chaincodeId: 'iotservice',
  });

  const timeout = setTimeout(() => fatal('timed out waiting for events'), 120 * 1000);
  await Promise.all([handleDeviceEvents(isb), handleServiceEvents(isb), handleRequestEvents(isb)]);
  clearTimeout(timeout);

  isb.close();
}

main();

File: tests/e2e/java/src/com/example/e2e/EventHub.java
package com.example.e2e;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.logging.Logger;
import org.hyperledger.fabric.client.CloseableIterator;
import org.nexus_lab.iot_service_blockchain.sdk.Device;
import org.nexus_lab.iot_service_blockchain.sdk.DeviceEvent;
import org.nexus_lab.iot_service_blockchain.sdk.Sdk;
import org.nexus_lab.iot_service_blockchain.sdk.SdkOptions;
import org.nexus_lab.iot_service_blockchain.sdk.Service;
import org.nexus_lab.iot_service_blockchain.sdk.ServiceEvent;
import org.nexus_lab.iot_service_blockchain.sdk.ServiceRequest;
import org.nexus_lab.iot_service_blockchain.sdk.ServiceRequestEvent;
import org.nexus_lab.iot_service_blockchain.sdk.ServiceResponse;

public class EventHub {
  private static final String ORG_ID = "Org2MSP";
  private static final String ORG_DOMAIN = "org2.example.com";
  private static final String USER_NAME = "User1@org2.example.com";
  private static final String PEER_NAME = "peer0.org2.example.com";
  private static final String PEER_ENDPOINT = "localhost:9051";

  private static final Logger LOGGER = Logger.getLogger("eventhub");

  public static void main(String[] args) throws Exception {
    String fabricRoot = System.getenv("FABRIC_ROOT");
    String[] credentials =
        Utils.getCredentials(
            fabricRoot == null ? "" : fabricRoot, ORG_DOMAIN, USER_NAME, PEER_NAME);

    SdkOptions options = new SdkOptions();
    options.setOrganizationId(ORG_ID);
    options.setCertificate(credentials[0]);
    options.setPrivateKey(credentials[1]);
    options.setGatewayPeerEndpoint(PEER_ENDPOINT);
    options.setGatewayPeerServerName(PEER_NAME);
    options.setGatewayPeerTLSCertificate(credentials[2]);
    options.setNetworkName("mychannel");
    options.setChaincodeId("iotservice");

    Sdk isb = new Sdk(options);

    ScheduledFuture<?> timeout =
        Executors.newSingleThreadScheduledExecutor()
            .schedule(
                () -> Utils.fatal(LOGGER, "timed out waiting for responses"),
                120,
                TimeUnit.SECONDS);

    ThreadPoolExecutor executor = (ThreadPoolExecutor) Executors.newFixedThreadPool(3);

    List<Callable<Object>> tasks = new ArrayList<>();
    tasks.add(Executors.callable(new DeviceEventHandler(isb)));
    tasks.add(Executors.callable(new ServiceEventHandler(isb)));
    tasks.add(Executors.callable(new RequestEventHandler(isb)));
    executor.invokeAll(tasks);

    executor.shutdown();
    timeout.cancel(true);

    isb.close();
    System.exit(0);
  }

  private static class DeviceEventHandler implements Runnable {
    private static final Logger LOGGER = Logger.getLogger("device_event_handler");

    private Sdk isb;

    DeviceEventHandler(Sdk isb) {
      this.isb = isb;
    }

    @Override
    public void run() {
      Utils.log(LOGGER, "Watching for device events");

      CloseableIterator<DeviceEvent> events = isb.getDeviceRegistry().registerEvent();
      Map<String, Integer> expected = new HashMap<>();
      expected.put("register", 1);
      expected.put("deregister", 1);
      Map<String, Integer> actual = new HashMap<>();

      events.forEachRemaining(
          (event) -> {
            actual.put(event.getAction(), actual.getOrDefault(event.getAction(), 0) + 1);

            if ("register".equals(event.getAction()) || "deregister".equals(event.getAction())) {
              Device device = (Device) event.getPayload();
              if (!event.getDeviceId().equals(device.getId())
                  || !event.getOrganizationId().equals(device.getOrganizationId())) {
                Utils.fatal(LOGGER, "device ID or organization ID mismatch");
              }
            }

            for (Map.Entry<String, Integer> entry : expected.entrySet()) {
              if (actual.get(entry.getKey()) != entry.getValue()) {
                return;
              }
            }

            events.close();
          });

      for (Map.Entry<String, Integer> entry : expected.entrySet()) {
        if (actual.get(entry.getKey()) != entry.getValue()) {
          Utils.fatal(
              LOGGER,
              "should have received " + entry.getValue() + " device " + entry.getKey() + " events");
        }
      }

      Utils.log(LOGGER, "Done watching for device events");
    }
  }

  private static class ServiceEventHandler implements Runnable {
    private static final Logger LOGGER = Logger.getLogger("service_event_handler");

    private Sdk isb;

    ServiceEventHandler(Sdk isb) {
      this.isb = isb;
    }

    @Override
    public void run() {
      Utils.log(LOGGER, "Watching for service events");

      CloseableIterator<ServiceEvent> events = isb.getServiceRegistry().registerEvent();
      Map<String, Integer> expected = new HashMap<>();
      expected.put("register", 2);
      expected.put("deregister", 1);
      Map<String, Integer> actual = new HashMap<>();

      events.forEachRemaining(
          (event) -> {
            actual.put(event.getAction(), actual.getOrDefault(event.getAction(), 0) + 1);

            if ("register".equals(event.getAction()) || "deregister".equals(event.getAction())) {
              Service service = (Service) event.getPayload();
              if (!event.getServiceName().equals(service.getName())
                  || !event.getDeviceId().equals(service.getDeviceId())
                  || !event.getOrganizationId().equals(service.getOrganizationId())) {
                Utils.fatal(
                    LOGGER,
                    "event and payload device ID, organization ID, or service name mismatch");
              }
            }

            for (Map.Entry<String, Integer> entry : expected.entrySet()) {
              if (actual.get(entry.getKey()) != entry.getValue()) {
                return;
              }
            }

            events.close();
          });

      for (Map.Entry<String, Integer> entry : expected.entrySet()) {
        if (actual.get(entry.getKey()) != entry.getValue()) {
          Utils.fatal(
              LOGGER,
              "should have received "
                  + entry.getValue()
                  + " service "
                  + entry.getKey()
                  + " events");
        }
      }

      Utils.log(LOGGER, "Done watching for service events");
    }
  }

  private static class RequestEventHandler implements Runnable {
    private static final Logger LOGGER = Logger.getLogger("request_event_handler");

    private Sdk isb;

    RequestEventHandler(Sdk isb) {
      this.isb = isb;
    }

    @Override
    public void run() {
      Utils.log(LOGGER, "Watching for service request events");

      CloseableIterator<ServiceRequestEvent> events = isb.getServiceBroker().registerEvent();
      Map<String, Integer> expected = new HashMap<>();
      expected.put("request", 2);
      expected.put("respond", 2);
      expected.put("remove", 1);
      Map<String, Integer> actual = new HashMap<>();

      events.forEachRemaining(
          (event) -> {
            actual.put(event.getAction(), actual.getOrDefault(event.getAction(), 0) + 1);

            switch (event.getAction()) {
              case "request":
                ServiceRequest request = (ServiceRequest) event.getPayload();
                if (!event.getRequestId().equals(request.getId())) {
                  Utils.fatal(LOGGER, "event and payload request ID mismatch");
                }
                break;
              case "respond":
                ServiceResponse response = (ServiceResponse) event.getPayload();
                if (!event.getRequestId().equals(response.getRequestId())) {
                  Utils.fatal(LOGGER, "event and payload request ID mismatch");
                }
                break;
              case "remove":
                if (!event.getRequestId().equals(event.getPayload())) {
                  Utils.fatal(LOGGER, "event and payload request ID mismatch");
                }
                break;
            }

            for (Map.Entry<String, Integer> entry : expected.entrySet()) {
              if (actual.get(entry.getKey()) != entry.getValue()) {
                return;
              }
            }

            events.close();
          });

      for (Map.Entry<String, Integer> entry : expected.entrySet()) {
        if (actual.get(entry.getKey()) != entry.getValue()) {
          Utils.fatal(
              LOGGER,
              "should have received "
                  + entry.getValue()
                  + " service request "
                  + entry.getKey()
                  + " events");
        }
      }

      Utils.log(LOGGER, "Done watching for service request events");
    }
  }
}
